chain/client/src/sync.rs

```diff
@@ -777,36 +867,43 @@ impl StateSync {
                 );
             }
             ShardSyncStatus::StateDownloadParts => {
-                for (i, download) in new_shard_sync_download.downloads.iter_mut().enumerate() {
-                    if download.run_me.load(Ordering::SeqCst) {
-                        let target = possible_targets
-                            [thread_rng().gen_range(0, possible_targets.len())]
-                        .clone();
-                        download.run_me.store(false, Ordering::SeqCst);
-                        download.state_requests_count += 1;
-                        download.last_target = Some(target.clone());
-                        let run_me = download.run_me.clone();
-                        actix::spawn(
-                            self.network_adapter
-                                .send(NetworkRequests::StateRequestPart {
-                                    shard_id,
-                                    sync_hash,
-                                    part_id: i as u64,
-                                    target: target.clone(),
-                                })
-                                .then(move |result| {
-                                    if let Ok(NetworkResponses::RouteNotFound) = result {
-                                        // Send a StateRequestPart on the next iteration
-                                        run_me.store(true, Ordering::SeqCst);
-                                    }
-                                    future::ready(())
-                                }),
-                        );
-                    }
+                let possible_targets_sampler =
+                    SamplerLimited::new(possible_targets, MAX_STATE_PART_REQUEST);
+
+                for ((part_id, download), target) in new_shard_sync_download
+                    .downloads
+                    .iter_mut()
+                    .enumerate()
+                    .filter(|(_, download)| download.run_me.load(Ordering::SeqCst))
+                    .zip(possible_targets_sampler)
+                {
```

Comment:
This is complicated part and it's hard to understand what's happening here. Could you please add some comments to explain?

Reply:
Added comment.